crates/turborepo-vercel-api-mock/src/lib.rs

```diff
@@ -82,7 +94,69 @@ pub async fn start_test_server(port: u16) -> Result<()> {
                     team_id: Some(EXPECTED_SSO_TEAM_ID.to_string()),
                 })
             }),
+        )
+        .route(
+            "/v8/artifacts/:hash",
+            put(
+                |Path(hash): Path<String>, headers: HeaderMap, mut body: BodyStream| async move {
+                    let root_path = tempdir.path();
+                    let file_path = root_path.join(&hash);
+                    let mut file = OpenOptions::new()
+                        .append(true)
+                        .create(true)
+                        .open(&file_path)
+                        .unwrap();
+
+                    let duration = headers
+                        .get("x-artifact-duration")
+                        .and_then(|header_value| header_value.to_str().ok())
+                        .and_then(|duration| duration.parse::<u64>().ok())
+                        .unwrap_or(0);
+
+                    let mut durations_map = durations.lock().await;
+                    durations_map.insert(hash.clone(), duration);
+
+                    while let Some(item) = body.next().await {
+                        let chunk = item.unwrap();
+                        file.write_all(&chunk).unwrap();
+                    }
+
+                    (StatusCode::CREATED, Json(hash))
+                },
+            ),
+        )
+        .route(
+            "/v8/artifacts/:hash",
+            get(|Path(hash): Path<String>| async move {
+                let root_path = tempdir2.path();
+                let file_path = root_path.join(&hash);
+                let buffer = std::fs::read(file_path).unwrap();
+                let duration = durations2.lock().await.get(&hash).cloned().unwrap_or(0);
+                let mut headers = HeaderMap::new();
+
+                headers.insert(
+                    "x-artifact-duration",
+                    HeaderValue::from_str(&duration.to_string()).unwrap(),
+                );
+
+                (headers, buffer)
+            }),
+        )
+        .route(
+            "/v8/artifacts/:hash",
+            head(|Path(hash): Path<String>| async move {
+                let duration = durations3.lock().await.get(&hash).cloned().unwrap_or(0);
+                let mut headers = HeaderMap::new();
+
+                headers.insert(
+                    "x-artifact-duration",
+                    HeaderValue::from_str(&duration.to_string()).unwrap(),
```

Comment:
A bit surprised that this much ceremony is required here.

> You require MOAR TRAITS.

Reply:
Yeah...this is the unfortunate reality of rust crates. Either very trait heavy or type heavy, either way it can get confusing.